### PR TITLE
EVG-1919: update idle timeout if there's been command updates

### PR DESCRIPTION
**Pull request by tychoish**
(I don't think this actually works because it'll get bumped for
heartbeats, which means there'll be no timeouts, but it's a start)

### CODE REVIEW

File: proto/background.go [hunk 1]
```diff
@@ -46,19 +45,25 @@ func (a *Agent) startHeartbeat(ctx context.Context, tc *taskContext, heartbeat c
...
+
 	defer timer.Stop()
 	for {
 		select {
 		case <-timer.C:
-			tc.logger.Execution().Error("Hit idle timeout")
-			close(idleTimeout)
-			return
+			// check the time of the last successful
+			// request against the API. if i'
```
Comment (bsamek) [inline proto/background.go]: Fix comment.

File: proto/background.go [hunk 2]
```diff
@@ -46,19 +45,25 @@ func (a *Agent) startHeartbeat(ctx context.Context, tc *taskContext, heartbeat c
...
+			if time.Now().Sub(a.comm.LastMessageAt()) > idleTimeoutInterval {
+				tc.logger.Execution().Error("Hit idle timeout")
+				close(idleTimeout)
+				return
+			}
+			a.checkIn(ctx, tc, tc.currentCommand, idleTimeoutInterval, resetIdleTimeout)
```
Comment (bsamek) [inline proto/background.go]: This should reset the timeout to `idleTimeoutInterval.Sub(time.Now().Sub(a.comm.LastMessageAt()))` (with error checking that it's not negative, or save time.Now() on 61).

File: proto/task.go [hunk 3]
```diff
@@ -112,10 +112,16 @@ func (a *Agent) checkIn(ctx context.Context, tc *taskContext, command model.Plug
...
-	tc.currentCommand = command
-	if idleTimeout != nil {
-		idleTimeout <- duration
+
+	if idleTimeout == nil {
+		return
+	}
+
+	select {
+	case idleTimeout <- duration:
```
Comment (bsamek) [inline proto/task.go]: What's the reason for this? If `idleTimeout` is closed, does this hit the default case?

File: rest/client/request.go [hunk 4]
```diff
@@ -103,6 +103,19 @@ func (c *communicatorImpl) createRequest(info requestInfo, data interface{}) (*h
 	return r, nil
 }
 
+func (c *communicatorImpl) bumpLastSentTime() {
```
Comment (bsamek) [inline rest/client/request.go]: Redundant with implementations in rest/client/client.go.

File: rest/client/sender.go [hunk 5]
```diff
@@ -18,16 +18,23 @@ const (
...
+	comm          Communicator
+	cancel        context.CancelFunc
+	pipe          chan message.Composer
+	lastBatch     chan struct{}
+	signalEnd     chan struct{}
+	updateTimeout bool
 	*send.Base
 }
 
+func newTimeoutLogSender(ctx context.Context, comm Communicator, channel string, taskData TaskData) send.Sender {
```
Comment (bsamek) [inline rest/client/sender.go]: Use this new constructor in rest/client/client.go.

File: proto/task.go (hunk 3 above)
Comment (tychoish) [reply proto/task.go]: yep, this is (a) canonical way to do non-blocking safe sends.

File: rest/client/sender.go (hunk 5 above)
Comment (tychoish) [reply rest/client/sender.go]: done

File: rest/client/request.go (hunk 4 above)
Comment (tychoish) [reply rest/client/request.go]: removed

File: proto/background.go (hunk 1 above)
Comment (tychoish) [reply proto/background.go]: done

File: proto/background.go (hunk 2 above)
Comment (tychoish) [reply proto/background.go]: updated

**Pull request merged**